### DISCUSSION TITLE
Templated control not setting defaults after utilizing binding converter.

### DISCUSSION THREAD

**Discussion by pmankow** [AvaloniaUI/Avalonia#14175] [Q&A] [answered]
I have a program that is using numericupdown controls to allow the user to input numbers. There are no spinners and I set increment, min, max, and rounding. I only want whole numbers so if a user enters 127.3 it displays 127. However, if I bind to the value and utilize it in the codebehind I get 127.3. In order to work around this I am using a binding converter which seems to be working well except for two new issues it has caused:

1) When initializing the program the defaultValue (12345) from the templated controls no longer works. Instead it uses the ConvertBack default value (0).

2) If I enter invalid data into the numericupdown (such as bbb or 23kj...), instead of reverting to the defaultValue or any other valid data that was previously in the box, it now reverts to either the ConvertBack default value (some of the time) or to empty/null (most of the time). (I think this is has to do with the timing in which everything is initialized on startup. For instance, any controls that I add dynamically during runtime always default to empty unless I give them valid data first.) Ideally it will initialize with the defaultVaule (12345) and always revert back to the last known good value instead of 0 or null/empty.
Update: If I change the value in the control to a valid number and then enter invalid data it will display the default value (0).

Note, I am creating many instances of the templated control.

Part of Templated Control:
```
<NumericUpDown Text="{TemplateBinding Property=MRegisterSetting, Converter={StaticResource IntToStringConverter}, Mode=TwoWay}"/>
````
Part of Templated Control Codebehind:
```
public static readonly StyledProperty<int> MRegisterSettingProperty = AvaloniaProperty.Register<ModbusControl, int>(nameof(MRegisterSetting), defaultValue: 12345);

public int MRegisterSetting
{
    get => GetValue(MRegisterSettingProperty);
    set => SetValue(MRegisterSettingProperty, value);
}
```
Converter:
```
{
    public class IntToStringConverter : IValueConverter
    {
        public object Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
            if (value is int intValue)
            {
                return intValue.ToString(culture);
            }

            return string.Empty;
        }
        public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
            if (value is string stringValue && int.TryParse(stringValue, NumberStyles.Any, culture, out int result))
            {
                return result;
            }

            // Return a default int value if conversion fails
            return 0;
        }
    }
}
```
Does anyone have any input? Thank you!

**Comment by pmankow** [comment, marked as answer]
Disregard. I just changed my converter to act on the Value and made it an IntToDecimalConverter and everything works now. The downside is that I have a conversion in two places being done (one for the view in the UI for the user, and the other in the converter for the codebehind). It would be nice if numericupdowns had a VisibleValue property or something like that.

**Reply by IanRawley** [reply]
They kind of do.  You can set a FormatString on the NumericUpDown which will determine what the user sees.
